rclcpp/doc/multi_threaded_executor_redesign.rst

```diff
@@ -0,0 +1,62 @@
+======================================
+:code:`MultiThreadedExecutor` Redesign
+======================================
+
+:code:`MultiThreadedExecutor` may attempt to execute one executable object multiple times. 
+This occurs because an executable object can be identified as ready in multiple threads before it has been executed, and each of these threads will then try to execute it.
+To fix this, a general solution should do the following:
+
+* Work for all the types of executable objects: timers, subscriptions, services, clients, and waitables
+* Avoid spurious wake ups (which can be costly)
+
+Summary of discussion so far
+----------------------------
+
+Broadly speaking, two approaches were discussed:
+
+1. **Handle no data in the executor**: If more than one thread is trying to execute the same object and that object is consumed after it is executed once, check for the executable before trying to execute it:
+
+    .. code-block:: c++
+
+        void execute() {
+          if(buffer->has_data()) {
+            auto data = buffer->consume_data();
+            run_any_callback(data);
+          } else {
+            // do nothing
+          }
+        }
+
+   :Advantages: 
+     * simplicity
+     * extends interface in a way that generalizes well for all other types of executable objects (i.e., timers)
+   :Disadvantages: 
+     * spurious wake up calls
+     * needs to be documented well so every class that executes :code:`AnyExecutable` objects is thread-safe and doesn't hide a race condition
+
+
+2. **Take the data**: When an executable is ready, take it and store it so that it can be executed when execute is called:
+
+    .. code-block:: c++
+
+        void take_data(std::shared_ptr<void> & data) {
+          data = buffer->consume_data();
+        }
+
+        void execute(std::shared_ptr<void> & data) {
+          run_any_callback(data_);
+        }
```

Comment:
My main concern is how data is type erased here.
I don't think that's a good idea for a user facing API.

I think it's a better idea to have something like:

```cpp
std::function<void()> get_handle() {
  return [data=buffer->consume_data()]() {
    run_any_callback(data);
  };
}
```

In that way, data is not type erased.

IIRC @wjwwood raised the concern that this approach forces entities to have a way of "handling" the data, and it would be nice in the future to be able to create, for example, a `Subscription` where you can directly take the data.

Though that would be nice, all our current API is based on callbacks.
i.e.: you need to actually pass a callback when creating a `Subscription`/etc.

If we want to add support to a WaitSet+take approach (which I think it would be nice for some users), we will actually need to split entities e.g. `Subscription` and `ExecutableSubscription`.

In that approach, I think that waitables actually don't need to have the concept of "taking data" neither.
You only need to be able to check if the waitable is ready, and then you can execute whatever code is needed.

----

Longer term:

I think we should separate the concept of Waitable/Condition from the concept of Executable.
The current design mixes both, and  they aren't necessarily the same.

I would like to write how I think that can be done, but I don't think that's the goal of this discussion.

Reply:
If we don't erase the type in the interface, I'm convinced that https://github.com/ros2/rclcpp/pull/1275 is a good idea (each time I read the PR I'm a bit confused, but I think I'm convinced now :joy:).
Without that, I still don't like the proposal much (sorry for being a bit hardheaded).

Maybe there's a different way of avoiding erasing the type, instead of my `get_handle` idea. I would like to hear of other alternatives!

Reply:
I've added the `get_handle` method in with an example. Let me know what you think.

Reply:
looks good

Reply:
I'm still convinced #1275 is _not_ the right solution, also sorry for being hardheaded, and I don't have an issue with the type erasure, since the root cause is that we have to deal with type erased data at the the `rcl`/`rmw` layer anyways.

If we added the "handle" which contains a type erased pointer, which the callback could re-type internally, I really don't see that issue with what was proposed here, w.r.t. the `std::shared_ptr<void> & data`.

Reply:
> If we added the "handle" which contains a type erased pointer, which the callback could re-type internally, I really don't see that issue with what was proposed here, w.r.t. the std::shared_ptr<void> & data.

The difference is that you don't expose to the user a `void *`.